File: backend/cut_schedule.py
"""The cut schedule shared by the montage and Dream cards, plus Dream's per-frame plan.

A *cut schedule* splits a composition's window into parts: the rising edges of a wired
`trigger` (through the card's own hysteresis threshold), unioned with hand-placed manual
breakpoints, minus individually disabled gate cuts. The montage plays one extract per
part; Dream generates under one prompt per part. Both cards carry the same four data
fields, so they share this code rather than the intent — a schedule that drifted between
them would be a bug nobody could see.

This is a LEAF module (numpy + nothing of ours but `graph_modulators._gate_curve`) so the
schedule can be imported and tested without dragging in the render DAG — the same reason
`VIDEO_PRODUCERS` lives in `graph_common`. The dag-aware wrapper that re-resolves a
trigger at the project's editing fps stays in `graph_render` as `_cut_frames`.

Frontend mirror: `frontend/src/lib/cutSchedule.ts`. The two MUST agree — the timeline
draws from one and the render generates from the other — which `tests/fixtures/
cut_schedule_cases.json` is there to enforce from both sides.
"""

from __future__ import annotations

import numpy as np

from .graph_modulators import _gate_curve

# A gap between two sung lines shorter than this is not a real silence — it is the
# padding `segment.align_lines` adds to every `t1` for readability, clamped at the next
# line's `t0`. Cutting on those would spawn one-frame "instrumental" parts between every
# pair of back-to-back lines.
MIN_GAP_S = 0.3


def lyric_cuts(
    lines,
    seg_start: float,
    fps: float,
    nframes: int,
    *,
    instrumental: bool = True,
    skip_unaligned: bool = False,
) -> list[tuple[int, bool]]:
    """Cut frames derived from aligned lyric lines → [(frame, is_gap_start), …].

    `is_gap_start` marks a cut that OPENS a silence (an instrumental stretch) rather than
    a sung line, so the caller can give those parts their own prompt.

    Lyric lines are SONG-ABSOLUTE (`sources_text.py`), while every other time on a
    scheduled card is composition-LOCAL — this is the one place the two meet, so the
    conversion happens here and nowhere else.

    Only real silences become gap cuts. `align_lines` pads every `t1` to at least
    `t0 + 1.7s` and then clamps it to the next line's `t0`, so two back-to-back lines have
    a ZERO-length gap: `t1` is a display duration, not the moment singing stopped. Cutting
    on every `t1` would therefore litter the schedule with one-frame parts.

    `skip_unaligned` drops lines whose timings were interpolated rather than heard
    (`aligned: false`) — their `t0` is arithmetic, so a cut there lands where nothing is
    sung.
    """
    rows = []
    for ln in lines or []:
        try:
            t0, t1 = float(ln.get("t0")), float(ln.get("t1"))
        except (TypeError, ValueError, AttributeError):
            continue
        if skip_unaligned and ln.get("aligned") is False:
            continue
        rows.append((t0, t1))
    rows.sort()

    out: list[tuple[int, bool]] = []
    for i, (t0, t1) in enumerate(rows):
        out.append((int(round((t0 - seg_start) * fps)), False))
        if not instrumental:
            continue
        # What bounds this line's trailing silence: the next line, or — for the last
        # one — the END OF THE WINDOW. Using `t1 + MIN_GAP_S` there would be both
        # arbitrary and float-fragile (4.3 - 4.0 == 0.29999…, so the final silence
        # vanished depending on the numbers).
        nxt = rows[i + 1][0] if i + 1 < len(rows) else seg_start + nframes / fps
        if nxt - t1 >= MIN_GAP_S:
            out.append((int(round((t1 - seg_start) * fps)), True))
    # Frame 0 is where part 0 already starts, and anything past the window never plays —
    # `effective_cuts` applies the same clamp to the gate and manual sources.
    return [(f, g) for f, g in out if 1 <= f < nframes]


def effective_cuts(
    trigger: np.ndarray,
    d: dict,
    fps: float,
    nframes: int,
    lyric: list[int] | None = None,
) -> list[int]:
    """The effective cut frames: GATE rises (the trigger through the card's built-in
    hysteresis threshold, exactly like the slideshow) minus the individually DISABLED
    ones, unioned with the MANUAL breakpoints and any LYRIC cuts — sorted, deduped at
    frame granularity, clamped inside (0, nframes).

    `lyric` is the frame list from `lyric_cuts` (Dream's "follow the lyrics" mode). It
    joins the union at the same point as the manual breakpoints and obeys the same
    `disabledCuts` suppression, so an unwanted line boundary is switched off with the
    machinery that already exists rather than a second mechanism.

    `disabledCuts` stores composition-LOCAL seconds; an entry suppresses ANY cut within
    HALF A FRAME of it — gate or manual — so the match is deterministic and a gate cut
    that MOVED (threshold edit) re-enables itself. Suppressing manuals too matters (v17):
    a manual breakpoint sharing a disabled gate cut's frame used to resurrect the cut the
    user just clicked off, while the timeline (where the gate mark wins the collision
    pixel) showed it silenced — the render cut where the UI said it wouldn't. The editor's
    gestures keep such data rare (disabling sweeps same-frame manuals, placing a manual
    clears a stale disable), but saved projects carry it. Manual breakpoints are local
    seconds too."""
    gate = _gate_curve(
        trigger, {"threshold": d.get("threshold", 0.5), "hysteresis": d.get("hysteresis", 0.1)}
    )
    rises = np.nonzero(np.diff(gate) > 0)[0] + 1  # frame index where each cut lands
    disabled = []
    for t in d.get("disabledCuts") or []:
        try:
            disabled.append(float(t) * fps)
        except (TypeError, ValueError):
            continue

    def silenced(frame: int) -> bool:
        return any(abs(frame - f) <= 0.5 for f in disabled)

    cuts = {int(r) for r in rises if not silenced(int(r))}
    for bp in d.get("manualBreakpoints") or []:
        try:
            f = int(round(float((bp or {}).get("t")) * fps))
        except (TypeError, ValueError):
            continue
        if not silenced(f):
            cuts.add(f)
    for f in lyric or []:
        if not silenced(int(f)):
            cuts.add(int(f))
    return sorted(c for c in cuts if 1 <= c < nframes)


def part_starts(cuts: list[int], spans: list[int]) -> list[int]:
    """Absolute start frame of each PLAYED part. Frame 0 always starts part 0; part k
    swallows `spans[k]` effective cuts before the next starts. Cuts beyond the parts are
    IGNORED — as is a part whose starting cut never arrives — so the last STARTED part
    HOLDS to the segment end. Part k is active on [starts[k], starts[k+1])."""
    starts = [0]
    consumed = 0
    for span in spans[:-1]:  # the last part never hands over — its span is moot
        consumed += span
        if consumed - 1 >= len(cuts):
            break  # not enough cuts left — the part that just played holds
        starts.append(int(cuts[consumed - 1]))
    return starts


def _clamped_fades(prompts: list[dict], starts: list[int], nframes: int, fps: float) -> list[tuple]:
    """(fadeIn, fadeOut) seconds per part, capped so `fadeIn + fadeOut <= the part's own
    duration`, both scaled down proportionally on overflow.

    This is not tidiness — it is exactly the condition under which two adjacent
    transitions cannot overlap. Transition k-1 -> k ends at `c_k + i_k`; transition
    k -> k+1 begins at `c_k + D_k - o_k`; they stay disjoint iff `i_k + o_k <= D_k`. So
    the clamp is what guarantees at most TWO prompts ever blend at once, which is in turn
    what lets the embedding lerp take a pair (imagegen._dream_embeds) and the cache key
    take a pair (dream_cache.canonical_prompts).
    """
    out = []
    for k in range(len(starts)):
        end = starts[k + 1] if k + 1 < len(starts) else nframes
        dur = max(0.0, (end - starts[k]) / fps)
        p = prompts[k] if k < len(prompts) else {}
        fi = max(0.0, float(p.get("fadeIn") or 0.0))
        fo = max(0.0, float(p.get("fadeOut") or 0.0))
        total = fi + fo
        if total > dur and total > 0:
            scale = dur / total
            fi, fo = fi * scale, fo * scale
        out.append((fi, fo))
    return out


def dream_plan(
    cuts: list[int],
    prompts: list[dict],
    fps: float,
    nframes: int,
    *,
    seed: int = 1,
    seed_mode: str = "gate",
    reseed_frames: list[int] | None = None,
    scale=None,
    keep=None,
    shape: float = 1.0,
) -> list[dict]:
    """One `{prompt_a, prompt_b, w, seed, scale, keep}` per frame — what
    `imagegen.dream_frames` consumes and `dream_cache` keys on.

    `prompts` is the card's ordered prompt list; each entry may carry `fadeIn`, `fadeOut`
    and `span`. `scale` is the resolved `control_scale` curve (per frame) or None;
    `keep` likewise, and it only reaches the pipe when a `video` input is wired (with no
    source clip there is nothing to seed from).
    """
    if not prompts:
        raise ValueError("dream_plan: no prompts")
    nframes = max(1, int(nframes))
    spans = [max(1, int(p.get("span") or 1)) for p in prompts]
    starts = part_starts(cuts, spans)
    fades = _clamped_fades(prompts, starts, nframes, fps)

    # Reseed events. In `gate` mode an unwired `reseed` port falls back to the CUT
    # schedule, so the natural default is "a fresh image family per prompt" with nothing
    # wired; wiring a separate signal re-rolls WITHIN a part.
    events = sorted(reseed_frames) if reseed_frames is not None else list(cuts)

    plan = []
    for f in range(nframes):
        k = 0
        while k + 1 < len(starts) and starts[k + 1] <= f:
            k += 1
        a = str(prompts[k].get("text") or "")
        b, w = None, 0.0

        # Am I inside the transition INTO this part, or the one OUT of it?
        t = f / fps
        if k > 0:  # incoming edge: [c - fadeOut(k-1), c + fadeIn(k)]
            c = starts[k] / fps
            o, i = fades[k - 1][1], fades[k][0]
            if t < c + i:
                a, b = str(prompts[k - 1].get("text") or ""), str(prompts[k].get("text") or "")
                w = _weight(t, c, o, i, shape)
        if k + 1 < len(starts):  # outgoing edge: [c - fadeOut(k), c + fadeIn(k+1)]
            c = starts[k + 1] / fps
            o, i = fades[k][1], fades[k + 1][0]
            if t >= c - o and o > 0:
                a, b = str(prompts[k].get("text") or ""), str(prompts[k + 1].get("text") or "")
                w = _weight(t, c, o, i, shape)

        plan.append(
            {
                "prompt_a": a,
                "prompt_b": b if w > 0 else None,
                "w": float(w),
                "seed": _seed_for(f, seed, seed_mode, events),
                "scale": 0.7 if scale is None else float(scale[min(f, len(scale) - 1)]),
                "keep": 0.1 if keep is None else float(keep[min(f, len(keep) - 1)]),
            }
        )
    return plan


def _shaped(u: float, shape: float) -> float:
    """Remap a linear fade fraction `u` through the card's fade SHAPE.

    `shape` 1.0 is the identity — a linear ramp, and the default. Above 1 the curve
    flattens around the midpoint, so the fade spends most of its duration near w = 0.5;
    below 1 it steepens there.

    Why this exists: the two models do not interpolate alike. SD-Turbo morphs steadily
    across the whole 0->1 sweep, but Z-Image is continuous-yet-STEEP — measured on the
    step-01 probe, essentially all of its visual change happens inside w in [0.42, 0.62].
    A linear ramp therefore spends ~80% of an HD fade window showing almost nothing and
    ~20% doing all the work, which reads as a soft cut rather than a dissolve. At shape 3
    about 58% of the window lands inside that band instead of 20%.

    It is a control rather than a per-model constant on purpose: the band was measured on
    ONE prompt pair, and nothing says it is the same for others. A hardcoded curve would
    be right on the probe and silently wrong elsewhere; a knob is honest about that.
    """
    if shape == 1.0 or u <= 0.0 or u >= 1.0:
        return u
    s = 2.0 * u - 1.0
    return 0.5 + 0.5 * (1.0 if s >= 0 else -1.0) * abs(s) ** shape


def _weight(t: float, c: float, o: float, i: float, shape: float = 1.0) -> float:
    """Blend weight at time `t` for a transition at cut time `c` with `o` seconds of
    lead-out and `i` seconds of lead-in. Both zero is a hard cut: 0 before `c`, 1 at and
    after it. Otherwise a ramp across [c-o, c+i] — linear at shape 1, so at the cut itself
    you are `o/(o+i)` of the way across."""
    if o + i <= 0:
        return 1.0 if t >= c else 0.0
    if t <= c - o:
        return 0.0
    if t >= c + i:
        return 1.0
    return _shaped((t - (c - o)) / (o + i), shape)


def _seed_for(f: int, seed: int, mode: str, events: list[int]) -> int:
    """Deterministic in every mode — non-determinism would poison the frame cache."""
    if mode == "frame":
        return int(seed) + f
    if mode == "gate":
        n = 0
        for e in events:
            if e <= f:
                n += 1
            else:
                break
        return int(seed) + n
    return int(seed)  # "fixed"

File: frontend/src/lib/cutSchedule.ts
// The cut schedule shared by the montage and Dream cards — the frontend mirror of
// backend `cut_schedule.py`. The two must agree or the strip/timeline preview lies
// about where the render will cut; `cut_schedule_cases.json` is read by BOTH suites
// so neither side can be "fixed" alone.
//
// A cut schedule splits a composition's window into parts. The montage plays one
// extract per part; Dream generates under one prompt per part.
//
// Effective cuts = GATE rises (the trigger through the card's built-in hysteresis
// threshold — `riseFrames`) minus the individually DISABLED ones, unioned with the
// MANUAL breakpoints; sorted, deduped at frame granularity, clamped inside
// (0, nframes). `disabledCuts`/breakpoint times are composition-LOCAL seconds; a
// disabled entry suppresses ANY cut within HALF A FRAME of it — gate or MANUAL
// (deterministic, and a gate cut that MOVED under a threshold edit re-enables
// itself). Manuals must obey too (v17): one sharing a disabled gate cut's frame
// used to resurrect the cut backend-side while this mirror said it was silenced.

import type { DreamPrompt, LyricLine, MontageData } from "./types";

// One mark for the breakpoints timeline: where, from which source, and whether a
// gate cut is currently disabled (drawn greyed/struck, still visible — provenance).
export interface CutMark {
  frame: number;
  source: "gate" | "manual" | "lyric";
  disabled: boolean;
  breakpointId?: string; // manual only — the row to move/delete
}

// A gap between two sung lines shorter than this is not a real silence — it is the
// padding `align_lines` adds to every `t1` for readability, clamped at the next line's
// `t0`. Mirror of backend `cut_schedule.MIN_GAP_S`.
export const MIN_GAP_S = 0.3;

export interface LyricCut {
  frame: number;
  gap: boolean; // this cut OPENS a silence rather than a sung line
}

// Cut frames derived from aligned lyric lines (Dream's "follow the lyrics"). Mirror of
// backend `cut_schedule.lyric_cuts` — see it for why only REAL silences become gap cuts
// and why the absolute→local conversion happens here and nowhere else.
export function lyricCuts(
  lines: LyricLine[] | undefined,
  segStart: number,
  fps: number,
  nframes: number,
  opts: { instrumental?: boolean; skipUnaligned?: boolean } = {}
): LyricCut[] {
  const instrumental = opts.instrumental !== false;
  const rows = (lines || [])
    .filter((l) => l && Number.isFinite(l.t0) && Number.isFinite(l.t1))
    .filter((l) => !(opts.skipUnaligned && l.aligned === false))
    .map((l) => [l.t0, l.t1] as [number, number])
    .sort((a, b) => a[0] - b[0]);

  const out: LyricCut[] = [];
  rows.forEach(([t0, t1], i) => {
    out.push({ frame: Math.round((t0 - segStart) * fps), gap: false });
    if (!instrumental) return;
    // What bounds this line's trailing silence: the next line, or — for the last one —
    // the END OF THE WINDOW.
    const next = i + 1 < rows.length ? rows[i + 1][0] : segStart + nframes / fps;
    if (next - t1 >= MIN_GAP_S) {
      out.push({ frame: Math.round((t1 - segStart) * fps), gap: true });
    }
  });
  return out.filter((c) => c.frame >= 1 && c.frame < nframes);
}

export function cutMarks(
  gateRises: number[],
  data: Pick<MontageData, "manualBreakpoints" | "disabledCuts">,
  fps: number,
  nframes: number,
  lyric: number[] = []
): CutMark[] {
  const half = 0.5;
  const disabledFrames = (data.disabledCuts || []).map((t) => t * fps);
  const marks: CutMark[] = [];
  const taken = new Set<number>();
  for (const r of gateRises) {
    if (r < 1 || r >= nframes) continue;
    marks.push({
      frame: r,
      source: "gate",
      disabled: disabledFrames.some((f) => Math.abs(r - f) <= half),
    });
    taken.add(r);
  }
  for (const bp of data.manualBreakpoints || []) {
    const f = Math.round(bp.t * fps);
    if (f < 1 || f >= nframes) continue;
    // Same-frame collision: the gate mark wins the pixel (provenance display); the
    // manual row still exists in the data and stays editable from the list.
    if (taken.has(f)) continue;
    marks.push({
      frame: f,
      source: "manual",
      // A disabled entry silences manuals too — mirror of _effective_cuts (v17).
      disabled: disabledFrames.some((df) => Math.abs(f - df) <= half),
      breakpointId: bp.id,
    });
    taken.add(f);
  }
  // Lyric cuts last: on a frame already claimed by a gate or manual mark the earlier
  // provenance wins the pixel, the same collision rule manual marks follow.
  for (const f of lyric) {
    if (f < 1 || f >= nframes || taken.has(f)) continue;
    marks.push({
      frame: f,
      source: "lyric",
      disabled: disabledFrames.some((df) => Math.abs(f - df) <= half),
    });
    taken.add(f);
  }
  return marks.sort((a, b) => a.frame - b.frame);
}

// The frames the render actually cuts at (enabled marks only, deduped).
export function effectiveCuts(
  gateRises: number[],
  data: Pick<MontageData, "manualBreakpoints" | "disabledCuts">,
  fps: number,
  nframes: number,
  lyric: number[] = []
): number[] {
  return cutMarks(gateRises, data, fps, nframes, lyric)
    .filter((m) => !m.disabled)
    .map((m) => m.frame);
}

// Absolute start frame of each PLAYED extract (mirrors backend `_montage_starts`):
// frame 0 always starts extract 0; extract k swallows spans[k] cuts before the next
// starts; an extract whose starting cut never arrives holds — the last STARTED one
// runs to the window end.
export function partStarts(cuts: number[], spans: number[]): number[] {
  const starts = [0];
  let consumed = 0;
  for (const span of spans.slice(0, -1)) {
    consumed += span;
    if (consumed - 1 >= cuts.length) break;
    starts.push(cuts[consumed - 1]);
  }
  return starts;
}

// ---- Dream: the per-frame plan (mirror of backend cut_schedule.dream_plan) ---------

export interface DreamPlanStep {
  prompt_a: string;
  prompt_b: string | null;
  w: number;
  seed: number;
}

// (fadeIn, fadeOut) seconds per part, capped so `fadeIn + fadeOut <= the part's own
// duration`, both scaled proportionally on overflow.
//
// This is the invariant, not tidiness: transition k-1→k ends at `c_k + i_k`, transition
// k→k+1 begins at `c_k + D_k - o_k`, and they stay disjoint iff `i_k + o_k <= D_k`. So
// the clamp is what guarantees at most TWO prompts ever blend at once — which is what
// lets the embedding lerp take a pair and the cache key take a pair. The timeline must
// draw the CLAMPED value: showing the requested one would be a lie about the render.
export function clampedFades(
  prompts: DreamPrompt[],
  starts: number[],
  nframes: number,
  fps: number
): { fadeIn: number; fadeOut: number; clamped: boolean }[] {
  return starts.map((s, k) => {
    const end = k + 1 < starts.length ? starts[k + 1] : nframes;
    const dur = Math.max(0, (end - s) / fps);
    const p = prompts[k] || ({} as DreamPrompt);
    const fadeIn = Math.max(0, p.fadeIn || 0);
    const fadeOut = Math.max(0, p.fadeOut || 0);
    const total = fadeIn + fadeOut;
    if (total > dur && total > 0) {
      const scale = dur / total;
      return { fadeIn: fadeIn * scale, fadeOut: fadeOut * scale, clamped: true };
    }
    return { fadeIn, fadeOut, clamped: false };
  });
}

// Remap a linear fade fraction through the card's fade SHAPE (mirror of backend
// `_shaped`). 1.0 is the identity — a linear ramp, and the default. Above 1 the curve
// flattens around the midpoint so the fade spends most of its duration near w = 0.5.
//
// It exists because the two models do not interpolate alike: SD-Turbo morphs steadily
// across the whole sweep, but Z-Image is continuous-yet-STEEP — measured on the step-01
// probe, essentially all its change happens inside w ∈ [0.42, 0.62], so a linear ramp
// spends ~80% of an HD fade showing nothing. At shape 3 about 58% of the window lands
// inside that band. A control rather than a per-model constant on purpose: the band was
// measured on ONE prompt pair, so a hardcoded curve would be right there and silently
// wrong elsewhere.
export function fadeShaped(u: number, shape: number): number {
  if (shape === 1 || u <= 0 || u >= 1) return u;
  const s = 2 * u - 1;
  return 0.5 + 0.5 * Math.sign(s) * Math.abs(s) ** shape;
}

// Blend weight at time `t` for a transition at cut time `c` with `o` seconds of lead-out
// and `i` of lead-in. Both zero is a hard cut. Otherwise a ramp across [c-o, c+i] —
// linear at shape 1, so at the cut itself you are `o/(o+i)` of the way across.
export function blendWeight(t: number, c: number, o: number, i: number, shape: number = 1): number {
  if (o + i <= 0) return t >= c ? 1 : 0;
  if (t <= c - o) return 0;
  if (t >= c + i) return 1;
  return fadeShaped((t - (c - o)) / (o + i), shape);
}

function seedFor(f: number, seed: number, mode: string, events: number[]): number {
  if (mode === "frame") return seed + f;
  if (mode === "gate") {
    let n = 0;
    for (const e of events) {
      if (e <= f) n++;
      else break;
    }
    return seed + n;
  }
  return seed;
}

// One step per frame. `reseedFrames` null means the `reseed` port is unwired, so `gate`
// mode falls back to the CUT schedule — a fresh image family per prompt with nothing
// wired; wiring a separate signal re-rolls WITHIN a part.
export function dreamPlan(
  cuts: number[],
  prompts: DreamPrompt[],
  fps: number,
  nframes: number,
  opts: {
    seed?: number;
    seedMode?: string;
    reseedFrames?: number[] | null;
    shape?: number;
  } = {}
): DreamPlanStep[] {
  if (!prompts.length) return [];
  const n = Math.max(1, Math.round(nframes));
  const spans = prompts.map((p) => Math.max(1, Math.round(p.span || 1)));
  const starts = partStarts(cuts, spans);
  const fades = clampedFades(prompts, starts, n, fps);
  const seed = opts.seed ?? 1;
  const mode = opts.seedMode ?? "gate";
  const shape = opts.shape ?? 1;
  const events = (opts.reseedFrames ?? cuts).slice().sort((a, b) => a - b);

  const plan: DreamPlanStep[] = [];
  for (let f = 0; f < n; f++) {
    let k = 0;
    while (k + 1 < starts.length && starts[k + 1] <= f) k++;
    let a = prompts[k]?.text || "";
    let b: string | null = null;
    let w = 0;
    const t = f / fps;
    if (k > 0) {
      // the tail of the transition INTO this part
      const c = starts[k] / fps;
      const o = fades[k - 1].fadeOut;
      const i = fades[k].fadeIn;
      if (t < c + i) {
        a = prompts[k - 1]?.text || "";
        b = prompts[k]?.text || "";
        w = blendWeight(t, c, o, i, shape);
      }
    }
    if (k + 1 < starts.length) {
      // the head of the transition OUT of it
      const c = starts[k + 1] / fps;
      const o = fades[k].fadeOut;
      const i = fades[k + 1].fadeIn;
      if (o > 0 && t >= c - o) {
        a = prompts[k]?.text || "";
        b = prompts[k + 1]?.text || "";
        w = blendWeight(t, c, o, i, shape);
      }
    }
    plan.push({ prompt_a: a, prompt_b: w > 0 ? b : null, w, seed: seedFor(f, seed, mode, events) });
  }
  return plan;
}
